### ISSUE TITLE
Some questions about refinement using ICVL dataset. 

### ISSUE THREAD

**Issue by Sakuralala** [moberweger/deep-prior#5]
Hi @moberweger ,
I read your paper carefully and I have some questions in refinement.
1.In the paper,you said:"We will refer to this architecture as ORRef, for Refinement with Overlapping Regions. **It uses as input several patches of different sizes but all centered on the joint location predicted by the first stage.**" But when I read the code in main_icvl_com_refine.py I can't find that how you did this,I tried saving the resized images,but I found that the patches are just centered on the original depth-map......
2.In the paper,you said:"To further improve the accuracy of the location estimates, **we iterate this refinement step several times,by centering the network on the location predicted at the previous iteration.**"And I can't find this in the code...
Can you give me some advice?Thank you!

**Comment by moberweger** [comment]
Hi @Sakuralala ,
The parts of the paper you are referring to are not included in this repo. This repo only includes the first stage, ie the Deep Prior without the refinement stage.

**Comment by Sakuralala** [comment]
Hi @moberweger ,
So how can I find the refinement stage?Can you tell me in which paper  I can find it？Thanks a lot!

**Comment by moberweger** [comment]
Hi @Sakuralala ,
I am not sure what you mean. The refinement stage is in our paper, but not in the code of this repo.

**Comment by Sakuralala** [comment]
Thanks @moberweger ,
Sorry that I misunderstood what you mean yesterday,I understood just now.So the code of refinement has't been released now,is that right?And the code main_icvl_com_refine.py is just a refinement on com,do  I understand correctly?
And ,can you give me some advice on  how to use the first stage's output as the refinement stage's input?I'm confused about that....
Thanks again!

**Comment by moberweger** [comment]
Hi @Sakuralala 
Yes, the main_icvl_com_refine.py is only for refinement of the com!
For the refinement of the joints, you can use the first stage to get the initial position of the joints in 3D. Project the 3D position to 2D and use this location to crop a small patch (64x64) centred on this 2D location. This patch is then used to predict an update on the 3D joint position that points towards the ground truth. You can apply this update to the initial position, get new 2D location, crop the patch, estimate the update, and so on...

**Comment by Sakuralala** [comment]
Hi @moberweger ,
Thanks for your patience.I'll have a try.
Thanks again!

**Comment by Azpril45** [comment]
Hi @moberweger ,
I read your paper  and I have some questions in refinement just as @Sakuralala .
I think this repo only includes the first stage without the refinement stage.
I also saw anther deep hand code based on Caffe which was published by @jsupancic .
I realized that it also only includes the first stage(deep prior) without the refinement stage.
So the code of refinement stage has not been released right?
Thank you !

**Comment by moberweger** [comment]
Hi @Azpril45 ,
Yes it is true what you say. This repo contains the original code, and the repo of @jsupancic  is a reimplementation. The refinement is not publicly released, but it should not take too much effort to implement it. At some point I probably add it to this repo, but for now I do not have time for this.

**Comment by Azpril45** [comment]
Hi @moberweger ,
Thank you for your patience.
I will take a try.

**Comment by Sakuralala** [comment]
Hi @moberweger ,sorry that I have one more question about the refine stage in you code.You said the refine stage uses the output of first stage,but in you repo I realized that you just used the coordinates of the ground truth....Is that right?And I can't find the part of **"You can apply this update to the initial position, get new 2D location, crop the patch, estimate the update, and so on..."**(you told me some days ago),so can you tell me what's my problem is?
Oh,sorry.The last question:
Why can't I get the same or approximative result when I train the net of first stage using your code,I got a pool result on it.....I just changed the dataset's path....Is there some tricks? Or some parameters to change?
Sorry to bother you again,hope you can give me some advice,thank you!

**Comment by moberweger** [comment]
Hi @Sakuralala ,
Again, there is no refinement stage implemented in this repo ;) I gave some details on how one can implement it, but it is not here.

> Why can't I get the same or approximative result when I train the net of first stage using your code,I got a pool result on it.....I just changed the dataset's path....Is there some tricks? Or some parameters to change?

What did you exactly do? Which dataset did you use? It should work out of the box. The results you should obtain with this code, should be at least as good as shown in the paper.

**Comment by Sakuralala** [comment]
Hi @moberweger ,I use the ICVL dataset,and I just changed the dataset's path,no more changed.And the as for 'floatX',I set it to float32,does this matter?I did't change anything except the dateset's path......I ran the training code and finally it output some pictures,but it's really pool compared with your result in your paper.

**Comment by moberweger** [comment]
Hi @Sakuralala ,
floatX=float32 is necessary for running on a GPU, so this is fine. 

I just ran the code from this repo, and I got an error of ~11mm for the ICVL dataset, without data augmentation, etc. So the code should be OK.

**Comment by Sakuralala** [comment]
Hi @moberweger ,I changed the gpu to gtx1080ti (I first used gtx 960m) and ran it again, and this time I got a good result,so maybe the gpu matters......Finally thanks for your patience.

**Comment by moberweger** [comment]
@Sakuralala , 
that is interesting. The training should be independent of the GPU. However, there might be slight differences due to the usage of cuDNN, since it can introduce some "randomness" (numerically) into the training process.
